arbor/util/spatial_tree.hpp

```diff
@@ -0,0 +1,184 @@
+    template <typename F>
+    inline void for_each(const F &func) const {
+        std::visit(
+            [&](auto &&arg) {
+                using arg_type = std::decay_t<decltype(arg)>;
+                if constexpr (std::is_same_v<arg_type, node_data>) {
+                    for (const auto &node: arg) { node.for_each(func); }
+                }
+                if constexpr (std::is_same_v<arg_type, leaf_data>) {
+                    for (const auto &d: arg) { func(d); }
+                }
+            },
+            data_);
+    }
+
+    // Iterate over all points within the given bounding box recursively.
+    // func must have signature `void func(const T&)`.
+    template <typename F>
+    inline void bounding_box_for_each(const point_type &box_min,
+        const point_type &box_max,
+        const F &func) const {
+        auto all_smaller_eq = [](const point_type &lhs, const point_type &rhs) {
+            bool result = true;
+            for (std::size_t i = 0; i < DIM; ++i) { result &= lhs[i] <= rhs[i]; }
+            return result;
+        };
+
+        std::visit(
+            [&](auto &&arg) {
+                using arg_type = std::decay_t<decltype(arg)>;
+
+                if (all_smaller_eq(box_min, min_) && all_smaller_eq(max_, box_max)) {
+                    // sub-nodes fully inside box -> call without further boundary
+                    // checks
+                    if constexpr (std::is_same_v<arg_type, node_data>) {
+                        for (const auto &node: arg) { node.template for_each<F>(func); }
+                    }
+                    if constexpr (std::is_same_v<arg_type, leaf_data>) {
+                        for (const auto &d: arg) { func(d); }
+                    }
+                }
+                else {
+                    // sub-nodes partially overlap bounding box
+                    if constexpr (std::is_same_v<arg_type, node_data>) {
+                        for (const auto &node: arg) {
+                            if (all_smaller_eq(node.min_, box_max) &&
+                                all_smaller_eq(box_min, node.max_))
+                                node.template bounding_box_for_each<F>(box_min, box_max, func);
+                        }
+                    }
+                    if constexpr (std::is_same_v<arg_type, leaf_data>) {
+                        for (const auto &d: arg) {
+                            const auto p = location_(d);
+                            if (all_smaller_eq(p, box_max) && all_smaller_eq(box_min, p)) {
+                                func(d);
+                            }
+                        }
+                    }
+                }
+            },
+            data_);
+    }
+
+    inline std::size_t size() const { return size_; }
+
+    inline bool empty() const { return !size_; }
+
+private:
+    std::size_t size_;
+    point_type min_, max_;
+    std::variant<node_data, leaf_data> data_;
```

Comment:
I've tried to measure the overhead, and found that for 10'000'000 calls, it only took an extra 30ms in total. However, I've added a utility function, which provides compile time checks to ensure all variant types are handled, but uses the index for case distinction. From my measurements, the performance is now incidental to a switch case statement.
This might be also helpful for your pull request, where you made similar changes.

Reply:
So, in earlier benchmarks removing `std::visit` from cell construction was a substantial speed-up. It might also depend on compiler (version).

Reply:
See also here:
https://stackoverflow.com/questions/57726401/stdvariant-vs-inheritance-vs-other-ways-performance